pow-migration: Update comments and clean up code

jsdanielh opened the pull request:
- Update comments and clean up code
  - Update comments and documentation.
  - Clean up code by using better code patterns and remove unnecessary indentation.
- Re-organize code for simplicity:
  - Removed duplicated types.
  - Moved all types to a single place.
  - Remove unnecessary directory nesting.

## Pull request checklist

- [X] All tests pass. The project builds and runs.
- [X] I have resolved any merge conflicts.
- [X] I have resolved all `clippy` and `rustfmt` warnings.

jsdanielh committed:
pow-migration: Update comments and clean up code

- Update comments and documentation.
- Clean up code by using better code patterns and remove unnecessary
  indentation.

viquezclaudio commented on pow-migration/README.md:
```diff
@@ -1,25 +1,42 @@
+  Nimiq PoW chain.
+- Validator readiness: Sends a validator ready transaction to signal that the validator
+  is ready to start the migration process and also monitors the PoW chain for other
+  validators readiness signal. When enough validators are ready, then a transition
```
This is not accurate. When enough validators are ready, we start the PoS chain. 
The transition block is candidate is pre-selected and hard-coded as part of this tool

viquezclaudio commented on pow-migration/README.md:
```diff
@@ -1,25 +1,42 @@
+  During this phase only pre-stake transactions are considered. By the end of this
+  phase, we would know the exact stake distribution of the registered validators.
+- Activation: This phase starts with a candidate transition block for PoS and ends
+  when validators confirm a candidate transition block. During this phase only transactions
```
ends when the 80% of the stake is ready

viquezclaudio commented on pow-migration/src/genesis/types.rs:
```diff
@@ -39,21 +39,22 @@ pub enum Error {
-/// 4. The final block in the PoW chain that will be taken as genesis block for the
-///    PoS chain. This block must have a block number that can be an election block
-///    number in the PoS chain.
+/// 3. Pre-stake registration window end block.
```
Point 2 needs to be updated, since the validator registration end block does not necessary equals the pre-stake start

jsdanielh replied:
In this code, sadly yes. I'll change this in my upcoming set of changes

viquezclaudio commented on pow-migration/src/genesis/types.rs:
```diff
@@ -39,21 +39,22 @@ pub enum Error {
 /// 5. This is a block whose block number is a number of confirmations away from
 ///    the final block described in 4.
 ///
+/// Note that after 5 there are some other blocks that are considered in the migration
```
I don't understand this note.
After 5 is when we start the genesis generation process

jsdanielh replied:
Removed it. It is going to be removed anyway in my next set of changes

viquezclaudio commented on pow-migration/src/lib.rs:
```diff
@@ -69,38 +69,38 @@ static MAINET_BLOCK_WINDOWS: &BlockWindows = &BlockWindows {
 
-    /// Number of confirmations after the final block needed for the PoS chain to
-    /// start.
+    /// Number of confirmations after any block to be considered it final in the PoW chain.
```
Number of confirmations after any block is considered final in the PoW

viquezclaudio commented on pow-migration/src/main.rs:
```diff
@@ -71,12 +72,11 @@ fn initialize_logging() {
-    //    4 - Monitor the PoW chain to detect if the genesis candidate is forked
-    //    5 - After X confirmations, start the 2.0 client with the generated genesis block
-    //    6 - If a fork is detected, go back to step 3 and repeat
+    //    1 - Wait for the pre-stake window to close (by waiting the confirmations).
```
Shouldn't we allow to pro-actively migrate history, even if the pre-stake window is not closed ?

jsdanielh replied:
Yes, code does it. Fixed comments

viquezclaudio commented on pow-migration/src/main.rs:
```diff
@@ -71,12 +72,11 @@ fn initialize_logging() {
-    //    5 - After X confirmations, start the 2.0 client with the generated genesis block
-    //    6 - If a fork is detected, go back to step 3 and repeat
+    //    1 - Wait for the pre-stake window to close (by waiting the confirmations).
+    //    2 - Select the next genesis candidate and wait until that block is mined
```
and this is the step where we generate the genesis

jsdanielh committed:
pow-migration: Re-organize code for simplicity

Re-organize code for simplicity:
- Removed duplicated types.
- Moved all types to a single place.
- Remove unnecessary directory nesting.

viquezclaudio commented on pow-migration/README.md:
```diff
@@ -1,25 +1,42 @@
-The intended users of these utilities are nodes who want to become validators in the Nimiq PoS chain.
+This repository contains a tool for the migration process from Nimiq Proof-of-Work
+to Nimiq Proof-of-Stake (PoS).
+The intended users of this utility are nodes who want to become validators in the
```
who want to become the *first* validators in the Nimiq PoS chain...

viquezclaudio commented on pow-migration/README.md:
```diff
@@ -1,25 +1,42 @@
+- Nimiq PoS state builder: Migrates the Nimiq PoW state into Nimiq PoS. This is a
+  requisite for the genesis block since its accounts tree root is calculated from
+  this state.
+- Migration binary: Starts the Nimiq PoS client when enough validators signaled a
```
Starts the Nimiq PoS client when enough validators are ready in the expected block window

viquezclaudio commented on pow-migration/src/types.rs:
```diff
@@ -0,0 +1,201 @@
+/// 5. The transition candidate block in the PoW chain that will be taken as genesis
+///    block for the PoS chain.
+/// 6. This is a block whose block number is a number of confirmations away from
+///    the final block described in 5 and marks the beginning of the activation window.
```
the beginning of the activation window is marked by the candidate block
